### ISSUE TITLE
[Unit Tests] Finish the MockServer to be used on tests

### ISSUE THREAD

**Issue by GPrimola** [GPrimola/jenkiexs#5]
The mock server is `Jenkiexs.MockServer.Jenkins`.
The idea is to mock calls to a real jenkins server.

Feel free to propose a different approach 👍

**Comment by sreecodeslayer** [comment]
From what I can tell from the Jobs api used by the httpoison client, wouldn't it be simpler to have a Bypass server in tests?

Ref: https://github.com/PSPDFKit-labs/bypass

We can stub and expect requests called and even assert the request params and headers

**Comment by GPrimola** [comment]
Hey @sreecodeslayer 

That's a great question!

**TL;DR;**

For the sake of this project's completeness, you can install Bypass and have it used in the tests, maybe my assumption don't apply for this project. :)

---

Explanation: (sorry whether this seems like a lecture!)
In my experience the answer is that it depends, and I really don't know if it'd be the best choice for this project.
I've used Bypass in some of http clients i've developed, and in some of them the stub function ends up being too big, with `case` pattern matches and stuff; besides, if you have double logic to generate access token and stuff, it gets even bigger.

In those clients I even use [ExMachina](https://github.com/thoughtbot/ex_machina) to have data outputs for every endpoint or request types, to test the client, so you can imagine how'd that be when having different logic for every GET request with different parameters.

At the beginning of a project it's tough to know whether Bypass is suitable enough for the case. That's why I decided to go with a mock server for this project, because I pictured it as being able to interact as much as possible with a Jenkins server through its API, as fine grained as possible, and that tends to need more logic on the test server.

But if you have a different experience, please share! Maybe it's our rush that led us on having a mess when using Bypass.

Nevertheless, thank you very much for your suggestion and interest!

**Comment by sreecodeslayer** [comment]
Not really sure having different params for request would get cumbersome in this case.

I get the point, but wouldn't the fake server also need to take care of different responses. Also you'd have bake in the logic of different response code handling into tests, or into the fake server which is almost like rebuilding another bypass. Only then would we be able to expect the patters you are referring to right? Cmiiw here or am missing something in the client that needs something more that bypass.


As for factories, all http clients would need expectation response factories defined somewhere 🤔 
Its just the matter of whether you nerd to have them as part of the module or just in test fixtures.

Fake servers makes sense when you have underlying otp behaviour like genserver calls and casts that needs to be asserted as part of tests. Do we have such cases in plan?

**Comment by GPrimola** [comment]
@sreecodeslayer I see your point and I agree with you! Of course mock servers has its drawbacks, some of them you've pointed.

> As for factories, all http clients would need expectation response factories defined somewhere

I lost you here. Could you be more clear?

> Fake servers makes sense when you have underlying otp behaviour like genserver calls and casts that needs to be asserted as part of tests.

Do you have any example of this? I just got curious about!

> Do we have such cases in plan?

Which cases are you referring to? If You mean test cases, do you think it'd be better I separate the issues by test case?

**Comment by GPrimola** [comment]
@sreecodeslayer ?

**Comment by sreecodeslayer** [comment]
oops, sorry i missed your last comment.

> I lost you here. Could you be more clear?

What i meant was, if a mock server is used, you need to have the actual jenkins response payload with dummy values given as the return value for the various endpoints you are trying to call. cmiiw
If this is the case, then wouldn't you be storing the response payload somewhere as part of the app?
You'd do the same with bypass as well?


> Do you have any example of this? I just got curious about!

I don't have any public example to show you, but I have worked on something similar. Let me try to explain the use-case. Think of an app that sends commands to many of it's agents [app on some other machine], via websockets/http. Now this communication is wrapped up in genserver so that the failed commands can be retried.
Such cases can be or part of it can be mocked if you are trying to test the genserver functionalities or retries.

> Which cases are you referring to? If You mean test cases, do you think it'd be better I separate the issues by test case?

The case I just said above, where you have genservers and other OTP behaviors 😄 

Sorry if i went too deep with this.

For mock server, how do you plan to build one?
Is that mock server going to be a real http server or you just mock the functions to return a valid response?

**Comment by sreecodeslayer** [comment]
Also, I was looking at the current MockServer. What we have there is basically a mini-Bypass 🤔 
I mean, we have Plug, we have actual routes and json responses stored as factories. Why not just bypass then? 😄
